### PR TITLE
gh-89550: Buffer GzipFile.write to reduce execution time by ~15%

### PR DESCRIPTION
**Pull request by CCLDArjun**
Currently, all of this is done for every write() call:
https://github.com/python/cpython/blob/main/Lib/gzip.py#L266-L288

This pr stores GzipFile.write() data into a buffer to postpone most of the compression work to when the buffer flushes. This way, we can call into zlib and `File.write()` less frequently.

benchmarking with timeit:
```py
../python -m timeit -s 'import gzip
' 'with gzip.open("./Python-3.11.1.tgz", "rb") as infile:
' '    with gzip.open("./Python-3.11.1new.tgz", "wb") as outfile:
' '        for line in infile:
' '            outfile.write(line)
' '
'
1 loop, best of 5: 6.36 sec per loop
```
vs
```py
../python -m timeit -s 'import gzipold as gzip
' 'with gzip.open("./Python-3.11.1.tgz", "rb") as infile:
' '    with gzip.open("./Python-3.11.1new.tgz", "wb") as outfile:
' '        for line in infile:
' '            outfile.write(line)
' '
'
1 loop, best of 5: 7.68 sec per loop
```

* Issue: gh-89550

### CODE REVIEW

**Comment by rhpvorderman** [comment]
Wow @CCLDArjun such a simple and effective solution. My solution was to split out a _GzipWriter (just like _GzipReader) and try to work from there, but I never got that to work easily and it was quite some extra code too. I am so glad you made this. I hope this will get merged soon!

**Comment by CCLDArjun** [comment]
> My solution was to split out a _GzipWriter (just like _GzipReader) and try to work from there

Initially that was my idea too, where I subclassed `_pyio.BufferedWriter` to override its flush methods but that also wasn't working easily and was pretty messy: https://github.com/python/cpython/pull/101251/commits/943ca9c92c9a897d657ea19ea7ed1e7d2f49a289

**Comment by rhpvorderman** [comment]
I am glad  that you were able to come up with a better idea. I have been hacking at the gzip module pretty thoroughly but I could not come up with this. I am in awe :grinning: . I will port this to [python-isal](https://github.com/pycompression/python-isal) and I have recently started on making a [zlib-ng module](https://github.com/pycompression/python-zlib-ng) too. That way people will not have to wait for 3.12 to ship to get all this fast compression.

**Comment by orsenthil** [comment]
This looks good to me. A core developer in the expert list will review before merging.

**Comment by rhpvorderman** [comment]
@orsenthil This is still labeled "awaiting review" rather then "awaiting core review" does that matter?

**Comment by arhadthedev** [comment]
@rhpvorderman Only review messages from the Python team count.

@Yhg1s, @gpshead as zipfile experts (not quite the same but still).

File: Lib/gzip.py
Comment (gpshead) [inline Lib/gzip.py]: This can create an unreasonably huge temporary bytes object.  Keep the existing logic and just change 1024 to something larger like 100_000.  Or perhaps even better, be consistent and assign the buffer_size I suggested passing above to a gzip module constant (global): `_WRITE_BUFFER_SIZE = 4*io.buffer_size` and use that constant here and above.

File: Lib/gzip.py
Comment (gpshead) [inline Lib/gzip.py]: Adding a `buffer_size=4*io.DEFAULT_BUFFER_SIZE` argument is probably worthwhile. The default is reasonable but not large.

File: Lib/gzip.py
Comment (gpshead) [inline Lib/gzip.py:302]: add a clarifying comment `# Called by our self._buffer underlying WriteBufferStream.`

**Review by gpshead — CHANGES_REQUESTED** [review]
Overall I like this change. It's great how cleanly this can be added. :)

**Comment by CCLDArjun** [comment]
I have made the requested changes; please review again

**Comment by CCLDArjun** [comment]
@gpshead tests are passing! Can you re review?

File: Lib/gzip.py
Comment (svisser) [inline Lib/gzip.py:251]: Should there be a comment to clarify the difference between `self.bufsize` and `self.buffer_size`?

Comment (CCLDArjun) [reply Lib/gzip.py:251]: Looks like `self.bufsize` is actually not used anywhere, at least in the gzip module. Might be best to just delete it?

**Comment by CCLDArjun** [comment]
pinging @gpshead for a final review

**Review by gpshead — APPROVED** [review]

**Pull request merged**
